File: js/lib/experimental/transport/domain/time-series-data/TimeSeriesData.ts
import { AssetIdentifier } from "../AssetIdentifier";
import { DataId } from "./DataId";

export namespace TimeSeries {
    export type TimeSeriesDataPackage = {
        package: Package;
    };

    export type Package = {
        header?: Header;
        timeSeriesData: TimeSeriesData[];
    };

    export type Header = {
        assetId: AssetIdentifier;
        timeSpan?: TimeSpan;
        dateCreated?: Date;
        dateModified?: Date;
        author?: string;
        systemConfiguration?: ConfigurationReference[];
        customHeaders?: { [k: string]: unknown };
    };

    export type ConfigurationReference = {
        id: string;
        timeStamp: Date;
    };

    export type TimeSpan = { start: Date; end: Date };

    export type TimeSeriesData = {
        dataConfiguration?: ConfigurationReference;
        tabularData?: TabularData[];
        eventData?: EventData;
        customProperties?: { [k: string]: unknown };
    };

    export type TabularData = {
        numberOfDataSet?: number;
        numberOfDataPoints?: number;
        dataId?: DataId[];
        dataSet?: TabularDataSet[];
    };

    export type TabularDataSet = {
        timeStamp: Date;
        value: string[];
        quality?: string[];
    };

    export type EventData = {
        numberOfDataSet?: number;
        dataSet?: EventDataSet[];
    };

    export type EventDataSet = {
        timeStamp: Date;
        dataId: DataId;
        value: string;
        quality?: string;
    };
}

File: csharp/src/Vista.SDK/Experimental/Transport/TimeSeriesData/TimeSeriesData.cs
namespace Vista.SDK.Experimental.Transport.TimeSeries;

public sealed record TimeSeriesDataPackage(Package Package);

public sealed record Package(Header? Header, IReadOnlyList<TimeSeriesData> TimeSeriesData);

public sealed record Header(
    AssetIdentifier AssetId,
    TimeSpan? TimeSpan,
    DateTimeOffset? DateCreated,
    DateTimeOffset? DateModified,
    string? Author,
    IReadOnlyList<ConfigurationReference>? SystemConfiguration,
    IReadOnlyDictionary<string, object> CustomHeaders
);

public sealed record ConfigurationReference(string Id, DateTimeOffset TimeStamp);

public sealed record TimeSpan(DateTimeOffset Start, DateTimeOffset End);

public sealed record TimeSeriesData(
    ConfigurationReference? DataConfiguration,
    IReadOnlyList<TabularData>? TabularData,
    EventData? EventData,
    IReadOnlyDictionary<string, object> CustomProperties
);

public sealed record TabularData(
    int? NumberOfDataSet,
    int? NumberOfDataPoints,
    IReadOnlyList<DataId>? DataId,
    IReadOnlyList<TabularDataSet>? DataSet
);

public sealed record TabularDataSet(
    DateTimeOffset TimeStamp,
    IReadOnlyList<string> Value,
    IReadOnlyList<string>? Quality
);

public sealed record EventData(int? NumberOfDataSet, IReadOnlyList<EventDataSet>? DataSet);

public sealed record EventDataSet(DateTimeOffset TimeStamp, DataId DataId, string Value, string? Quality);
